src/range.rs

```diff
@@ -7,354 +7,458 @@
+                Self::negate_segments(Included(v.clone()), &self.segments[1..])
             }
+            Some((Included(_), Included(_)))
+            | Some((Included(_), Excluded(_)))
+            | Some((Excluded(_), Included(_)))
+            | Some((Excluded(_), Excluded(_))) => Self::negate_segments(Unbounded, &self.segments),
         }
     }
 
     /// Helper function performing the negation of intervals in segments.
-    /// For example:
-    ///    [ (v1, None) ] => [ (start, Some(v1)) ]
-    ///    [ (v1, Some(v2)) ] => [ (start, Some(v1)), (v2, None) ]
-    fn negate_segments(start: V, segments: &[Interval<V>]) -> Range<V> {
-        let mut complement_segments = SmallVec::empty();
-        let mut start = Some(start);
-        for (v1, maybe_v2) in segments {
-            // start.unwrap() is fine because `segments` is not exposed,
-            // and our usage guaranties that only the last segment may contain a None.
-            complement_segments.push((start.unwrap(), Some(v1.to_owned())));
-            start = maybe_v2.to_owned();
-        }
-        if let Some(last) = start {
-            complement_segments.push((last, None));
+    fn negate_segments(start: Bound<V>, segments: &[Interval<V>]) -> Self {
+        let mut complement_segments: SmallVec<Interval<V>> = SmallVec::empty();
+        let mut start = start;
+        for (v1, v2) in segments {
+            complement_segments.push((
+                start,
+                match v1 {
+                    Included(v) => Excluded(v.clone()),
+                    Excluded(v) => Included(v.clone()),
+                    Unbounded => unreachable!(),
+                },
+            ));
+            start = match v2 {
+                Included(v) => Excluded(v.clone()),
+                Excluded(v) => Included(v.clone()),
+                Unbounded => Unbounded,
+            }
+        }
+        if !matches!(start, Unbounded) {
+            complement_segments.push((start, Unbounded));
         }
 
         Self {
             segments: complement_segments,
         }
     }
+}
+
+impl<V: Ord> Range<V> {
+    /// Convert to something that can be used with
+    /// [BTreeMap::range](std::collections::BTreeMap::range).
+    /// All versions contained in self, will be in the output,
+    /// but there may be versions in the output that are not contained in self.
+    /// Returns None if the range is empty.
+    pub fn bounding_range(&self) -> Option<(Bound<&V>, Bound<&V>)> {
+        self.segments.first().map(|(start, _)| {
+            let end = self
+                .segments
+                .last()
+                .expect("if there is a first element, there must be a last element");
+            (bound_as_ref(start), bound_as_ref(&end.1))
+        })
+    }
+
+    /// Returns true if the this Range contains the specified value.
+    pub fn contains(&self, v: &V) -> bool {
+        if let Some(bounding_range) = self.bounding_range() {
```

Comment:
We should benchmark if this pulls its weight. Theoretically it should, an `O(1)` check vs an `O(segments.len)` check. But in practice `segments.len` tends to be short so it may not be faster.

Reply:
I'll run the benchmarks with and without this code and get back to you!

Reply:
Speaking of which, do you maybe have the large dataset files so I can bench with those too?

Reply:
We have bean using the elm and zuse files from https://github.com/pubgrub-rs/pubgrub/issues/44
I have larger examples but so far they are too large to run through criterion or get meaningful results from.
By the way collecting benchmarks is trickier than it seems, some discussion at https://rust-lang.zulipchat.com/#narrow/stream/260232-t-cargo.2FPubGrub/topic/random.20bench.20perf.20changes

We should set up a more permanent place for sharing these test files.

Reply:
I did a bunch of profiling with early out at the start of the function, exiting the loop if ever `v < start`, or no optimization at all. The performance of all of these cases was worse than not checking at all (all within 2%-6%).